pgjdbc/src/main/java/org/postgresql/jdbc/PgClob.java

```diff
@@ -5,62 +5,153 @@
+    super(conn);
+  }
 
-  public PgClob(org.postgresql.core.BaseConnection conn, long oid) throws java.sql.SQLException {
+  PgClob(org.postgresql.core.BaseConnection conn, long oid) throws java.sql.SQLException {
     super(conn, oid);
   }
 
-  public synchronized Reader getCharacterStream(long pos, long length) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "getCharacterStream(long, long)");
+  private int getBufferSize(long length) {
+    return (int) Math.min(length * 1.2, 64000);
   }
 
+  @Override
   public synchronized int setString(long pos, String str) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "setString(long,str)");
+    // TODO: pos>0 check
+    return setString(pos, str, 0, str.length());
+  }
+
+  private long measureBytes(long numCharacters) throws SQLException {
+    if (numCharacters == 0) {
+      return 0;
+    }
+    // TODO: offload to backend somehow?
+    Reader characterStream = getCharacterStream(1, numCharacters);
+    int bufferSize = (int) (numCharacters < 8192 ? numCharacters * 1.2 : numCharacters);
+    char[] tmp = new char[bufferSize];
+    int read;
+    long len = 0;
+    try {
+      while ((read = characterStream.read(tmp)) != -1) {
+        len = len + read;
+      }
+    } catch (IOException e) {
+      throw new PSQLException(GT.tr("Unable to skip {0} characters, large object oid is {1}",
+          numCharacters, getOid()),
+          PSQLState.UNEXPECTED_ERROR, e);
+    }
+    return len;
   }
 
+
+  @Override
   public synchronized int setString(long pos, String str, int offset, int len) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "setString(long,String,int,int)");
+    // TODO: add pos>0 check and error message
+    Writer writer = setCharacterStream(pos);
+    try {
+      writer.write(str, offset, len);
+      return len;
+    } catch (IOException e) {
+      throw new SQLException();
+    } finally {
+      try {
+        writer.close();
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+
+    }
   }
 
+  @Override
   public synchronized java.io.OutputStream setAsciiStream(long pos) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "setAsciiStream(long)");
+    return setBinaryStream(pos);
   }
 
+  @Override
   public synchronized java.io.Writer setCharacterStream(long pos) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "setCharacteStream(long)");
+    Charset connectionCharset = Charset.forName(conn.getEncoding().name());
+    return new OutputStreamWriter(setBinaryStream(measureBytes(pos - 1) + 1), connectionCharset);
   }
 
+  @Override
   public synchronized InputStream getAsciiStream() throws SQLException {
     return getBinaryStream();
```

Comment:
This returns the binary stream of the clob in backend encoding, right?
Presuming the backend is utf-8, this could have multi byte characters. What should this method return? I